### PR TITLE
SAK-45674 - Calendar: Replace UI by fullcalendar

### PR DESCRIPTION
**Pull request by mpellicer**

### CODE REVIEW

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    initialView: 'timeGridWeek',
+    displayEventTime: false,
+    headerToolbar: {
+      left: 'prev,next today',
+      center: 'title',
+      right: 'dayGridMonth,timeGridWeek,timeGridDay,listWeek'
+    },
+    eventSources: [{
+      events: function(event, successCallback, failureCallback) {
+        var start_date = moment(event.start).format('YYYY-MM-DD');
```
Comment (adrianfish) [inline calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:53]: I wouldn't be mixing casing of my variables. You've got start_date, startDate, end_date, endDate. If you're only using start_date and end_date in requestData, why not just init them in the object literal? Your variables are a bit confusing.

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        var requestData = {
+          merged: true,
+          firstDate: start_date,
+          lastDate: end_date
+        };
+        // Fetch the site calendar events from the REST endpoint.
+        var calendarRequestUrl = new URL(`${sakaiOrigin}/direct/calendar/site/${siteId}.json`);
+        Object.keys(requestData).forEach(key => calendarRequestUrl.searchParams.append(key, requestData[key]));
+        window.fetch(calendarRequestUrl, requestInit)
+        .then(response => response.json())
```
Comment (adrianfish) [inline calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:64]: You need to handle network errors in here.
r => {
  if (r.ok) return r.json();
  else throw new Error("Network error while fetching calendar data");
})
...
.catch(error => console.error(error));

or something like that..

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        .then(json => {
```
Comment (adrianfish) [inline calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:65]: This isn't json anymore. No big deal, but it's now an object. I'd just call it data, or something else.

File: lessonbuilder/tool/src/webapp/templates/ShowPage.html [hunk 4]
```diff
@@ -65,12 +65,9 @@
...
  <link rel="stylesheet" href="/library/editor/ckextraplugins/folder-listing/css/file-tree.css" type="text/css"/>
- <link href="/library/webjars/fullcalendar/4.3.1/core/main.min.css" rel="stylesheet" />
- <link href="/library/webjars/fullcalendar/4.3.1/daygrid/main.min.css" rel="stylesheet" />
- <link href="/library/webjars/fullcalendar/4.3.1/timegrid/main.min.css" rel="stylesheet" />
- <script src="/library/webjars/fullcalendar/4.3.1/core/main.min.js"></script>
- <script src="/library/webjars/fullcalendar/4.3.1/daygrid/main.min.js"></script>
- <script src="/library/webjars/fullcalendar/4.3.1/timegrid/main.min.js"></script>
+ <link href="/library/webjars/fullcalendar/5.8.0/main.min.css" rel="stylesheet" />
+ <script src="/library/webjars/fullcalendar/5.8.0/main.min.js"></script>
+ <script src="/library/webjars/fullcalendar/5.8.0/locales-all.min.js"></script>
```
Comment (adrianfish) [inline lessonbuilder/tool/src/webapp/templates/ShowPage.html:70]: Should we have the git sha on the end of these urls? Cache busting?

**Review by adrianfish — CHANGES_REQUESTED** [review]
js stuff, const instead of var, variable name casing. Just making stuff a bit clearer.

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  <div id="calendarDiv"></div>
+  ## include the Legend:
+  #calendarLegend()
+</div>
+
+<link href="/library/webjars/fullcalendar/5.8.0/main.min.css" rel="stylesheet" />
+<script src="/library/webjars/fullcalendar/5.8.0/main.min.js"></script>
+<script src="/library/webjars/fullcalendar/5.8.0/locales-all.min.js"></script>
+<script>
+  var userLanguage = portal.locale || navigator.language;
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:25]: Okay.

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  var computedStyle = getComputedStyle(document.querySelector(':root'));
+  var eventBackgroundColor = computedStyle.getPropertyValue('--infoBanner-bgcolor', '#e9f5fc');
+  var eventTextColor = computedStyle.getPropertyValue('--infoBanner-color', '#196390');
+  var sakaiOrigin = window.location.origin;
+  var siteId = portal.siteId;
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:31]: Okay.

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+      return {
+        html: eventMainFrame.innerHTML
+      };
+    }
+  });
+  if (userLanguage) {
+    calendar.setOption('locale', userLanguage);
+  }
+  calendar.render();
+</script>
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:132]: Okay.

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+      // Append our custom event with icon to the default event structure.
+      eventTitle.appendChild(iconSpan);
+      eventTitle.appendChild(document.createTextNode(`${dollar}{title}`));
+      // Return the modified event including out event structure.
+      return {
+        html: eventMainFrame.innerHTML
+      };
+    }
+  });
+  if (userLanguage) {
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:128]: Okay

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm (hunk 1 above)
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:52]: Okay

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm (hunk 3 above)
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:64]: Okay, responseData.

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm (hunk 2 above)
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm:63]: Yeah that makes sense.

File: lessonbuilder/tool/src/webapp/templates/ShowPage.html (hunk 4 above)
Comment (mpellicer) [reply lessonbuilder/tool/src/webapp/templates/ShowPage.html:70]: Could you elaborate?

File: calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  ## include the Legend:
+  #calendarLegend()
+</div>
+
+<link href="/library/webjars/fullcalendar/5.8.0/main.min.css" rel="stylesheet" />
+<script src="/library/webjars/fullcalendar/5.8.0/main.min.js"></script>
+<script src="/library/webjars/fullcalendar/5.8.0/locales-all.min.js"></script>
+<script>
+  var userLanguage = portal.locale || navigator.language;
+  var computedStyle = getComputedStyle(document.querySelector(':root'));
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/vm/calendar/chef_calendar_view.vm]: Sure.

**Comment by mpellicer** [comment]
Re-tested after all of the requested changes, works fine in Chrome, Firefox and Edge.

**Review by profmikegreene — CHANGES_REQUESTED** [review]
I'd like to see `library/src/morpheus-master/sass/modules/tool/calendar/_calendar_variables.scss` updated to include values from
https://github.com/fullcalendar/fullcalendar/blob/master/packages/common/src/styles/vars.css
https://github.com/fullcalendar/fullcalendar/blob/master/packages/daygrid/src/styles/vars.css
https://github.com/fullcalendar/fullcalendar/blob/master/packages/list/src/styles/vars.css

to ensure fullcalendar uses styles defined by library.

File: calendar/calendar-tool/tool/src/webapp/js/sakai-calendar.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
+  const userLanguage = window.top.portal.locale || navigator.language;
+  // Get the event color from the skin variables.
+  const computedStyle = getComputedStyle(document.querySelector(':root'));
+  const eventBackgroundColor = computedStyle.getPropertyValue('--infoBanner-bgcolor', '#e9f5fc');
+  const eventTextColor = computedStyle.getPropertyValue('--infoBanner-color', '#196390');
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/js/sakai-calendar.js:5]: Could you guide me? I've been out for over a year and a half, didn't know that these styles and variables exist.

fullcalendar accepts these properties, I thought was a good idea to get the vars from the skin!.

[the remark this answers is not in the document]
Comment (mpellicer) [reply calendar/calendar-tool/tool/src/webapp/js/sakai-calendar.js:5]: Maybe we can merge this and continue styling in future PRs, not sure @profmikegreene , thanks for your help

File: lessonbuilder/tool/src/webapp/templates/ShowPage.html (hunk 4 above)
Comment (adrianfish) [reply lessonbuilder/tool/src/webapp/templates/ShowPage.html:70]: actually, I don't think it matters. You're using a versioned js anyway. Ignore me :)

File: calendar/calendar-tool/tool/src/webapp/js/sakai-calendar.js [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        const calendarRequestUrl = new URL(`${sakaiOrigin}/direct/calendar/site/${siteId}.json`);
+        Object.keys(requestData).forEach(key => calendarRequestUrl.searchParams.append(key, requestData[key]));
+        window.fetch(calendarRequestUrl, requestInit)
+        .then(response => {
+          if (response.ok) {
+            return response.json()
+          } else throw new Error("Network error while fetching calendar data");
+        })
+        .then(responseData => {
+          let events = [];
```
Comment (adrianfish) [inline calendar/calendar-tool/tool/src/webapp/js/sakai-calendar.js:45]:
```suggestion
          const events = [];
```

**Review by adrianfish — APPROVED** [review]
Looks good. I tried it out and it all seems to mesh fine with the admin parts.

**Pull request merged**